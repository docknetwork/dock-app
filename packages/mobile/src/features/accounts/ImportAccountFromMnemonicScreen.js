import {
  Button,
  FormControl,
  Stack,
  Tooltip,
  Pressable,
  Popover,
  TextArea,
} from 'native-base';
import React, {useState} from 'react';
import {useDispatch} from 'react-redux';
import {AccountAdvancedOptions} from '../../components/AccountAdvancedOptions';
import {
  Header,
  Footer,
  Content,
  ScreenContainer,
  Typography,
  SelectToggler,
  NBox as Box,
  Select,
  Input,
  InputPopover,
  LoadingButton,
} from '../../design-system';
import {BackButton} from '../../design-system/buttons';
import {translate} from '../../locales';
import {createAccountOperations} from '../account-creation/create-account-slice';

export function ImportAccountFromMnemonicScreen({
  form,
  onChange,
  submitDisabled,
  onSubmit,
}) {
  return (
    <ScreenContainer testID="create-wallet-screen">
      <Header>
        <BackButton />
      </Header>
      <Content marginLeft={26} marginRight={26}>
        <Typography variant="h1" marginTop={52}>
          {translate('import_account_from_mnemonic.title')}
        </Typography>
        <Typography marginTop={12}>
          {translate('import_account_from_mnemonic.description')}
        </Typography>
        <Box>
          <FormControl isInvalid={form._errors.phrase}>
            <Stack mt={7}>
              <FormControl.Label>
                {translate('import_account_from_mnemonic.phrase_input')}
              </FormControl.Label>
              <TextArea
                placeholder=""
                value={form.phrase}
                onChangeText={onChange('phrase')}
                autoCapitalize="none"
              />
              <FormControl.ErrorMessage>
                {translate('import_account_from_mnemonic.invalid_phrase')}
              </FormControl.ErrorMessage>
            </Stack>
          </FormControl>
        </Box>
        <AccountAdvancedOptions onChange={onChange} form={form} />
      </Content>
      <Footer marginBottom={10} marginLeft={26} marginRight={26}>
        <LoadingButton
          full
          testID="next-btn"
          isDisabled={submitDisabled}
          onPress={onSubmit}>
          {translate('navigation.next')}
        </LoadingButton>
      </Footer>
    </ScreenContainer>
  );
}

export function ImportAccountFromMnemonicContainer() {
  const dispatch = useDispatch();
  const [form, setForm] = useState({
    phrase: '',
    keypairType: 'sr25519',
    derivationPath: '',
    _errors: {},
    _hasError: false,
  });

  const handleChange = key => evt => {
    setForm(v => ({
      ...v,
      [key]: evt,
    }));
  };

  const handleSubmit = () => {
    return dispatch(createAccountOperations.importFromMnemonic(form));
  };

  const submitDisabled = !form.phrase;

  return (
    <ImportAccountFromMnemonicScreen
      form={form}
      submitDisabled={submitDisabled}
      onChange={handleChange}
      onSubmit={handleSubmit}
    />
  );
}
