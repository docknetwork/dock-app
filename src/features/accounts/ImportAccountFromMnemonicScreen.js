import {
  Button,
  FormControl,
  Stack,
  Tooltip,
  Pressable,
  Popover,
  TextArea,
} from 'native-base';
import React, {useState} from 'react';
import {useDispatch} from 'react-redux';
import {
  Header,
  Footer,
  Content,
  ScreenContainer,
  Typography,
  SelectToggler,
  NBox as Box,
  Select,
  Input,
  InputPopover,
  LoadingButton,
} from '../../design-system';
import {BackButton} from '../../design-system/buttons';
import {
  createAccountOperations,
} from '../account-creation/create-account-slice';

export function ImportAccountFromMnemonicScreen({
  form,
  onChange,
  submitDisabled,
  onSubmit,
}) {
  return (
    <ScreenContainer testID="create-wallet-screen">
      <Header>
        <BackButton />
      </Header>
      <Content marginLeft={26} marginRight={26}>
        <Typography
          fontFamily="Montserrat"
          fontSize={24}
          lineHeight={32}
          fontWeight="600"
          color="#fff"
          marginTop={52}>
          Account recovery phrase
        </Typography>
        <Typography
          fontSize={16}
          lineHeight={24}
          fontWeight="400"
          marginTop={12}>
          This is the 12 word phrase you were given when you created your account.
        </Typography>
        <Box>
          <FormControl isInvalid={form._errors.phrase}>
            <Stack mt={7}>
              <FormControl.Label>
                Enter account recovery phrase
              </FormControl.Label>
              <TextArea
                placeholder=""
                value={form.phrase}
                onChangeText={onChange('phrase')}
                autoCapitalize="none"
              />
              <FormControl.ErrorMessage>
                Invalid recovery phrase
              </FormControl.ErrorMessage>
            </Stack>
          </FormControl>
        </Box>

        <SelectToggler placeholder="Advanced options">
          <FormControl>
            <Stack mt={7}>
              <FormControl.Label>
                Keypair crypto type
                <InputPopover>
                  Determines what cryptography will be used to create this
                  account. Note that to validate on Polkadot, the session
                  account must use "ed25519".
                </InputPopover>
              </FormControl.Label>
              <Select
                onValueChange={onChange('keypairType')}
                selectedValue={form.keypairType}>
                <Select.Item
                  label="Schnorrkel (sr25519, recommended)"
                  value="sr25519"
                />
                <Select.Item
                  label="Edwards (ed25519, alternative)"
                  value="ed25519"
                />
                <Select.Item
                  label="ECDSA (Non BTC/ETH compatible)"
                  value="ecdsa"
                />
              </Select>
            </Stack>
          </FormControl>
          <FormControl isInvalid={false}>
            <Stack mt={7}>
              <FormControl.Label>
                Secret derivation path
                <InputPopover>
                  {`You can set a custom derivation path for this account using the following syntax "/<soft-key>//<hard-key>". The "/<soft-key>" and "//<hard-key>" may be repeated and mixed. An optional "///<password>" can be used with a mnemonic seed, and may only be specified once.`}
                </InputPopover>
              </FormControl.Label>
              <Input
                placeholder="//hard/soft//password"
                onChangeText={onChange('derivationPath')}
                autoCapitalize="none"
              />
              <FormControl.ErrorMessage>
                Unable to match provided value to a secret URI.
              </FormControl.ErrorMessage>
            </Stack>
          </FormControl>
        </SelectToggler>
      </Content>
      <Footer marginBottom={10} marginLeft={26} marginRight={26}>
        <LoadingButton
          full
          testID="next-btn"
          isDisabled={submitDisabled}
          onPress={onSubmit}>
          Next
        </LoadingButton>
      </Footer>
    </ScreenContainer>
  );
}

export function ImportAccountFromMnemonicContainer() {
  const dispatch = useDispatch();
  const [form, setForm] = useState({
    phrase: '',
    keypairType: 'sr25519',
    derivationPath: '',
    _errors: {},
    _hasError: false,
  });

  const handleChange = key => evt => {
    setForm(v => ({
      ...v,
      [key]: evt,
    }));
  };

  const handleSubmit = () => {
    return dispatch(createAccountOperations.importFromMnemonic(form));
  };

  const submitDisabled = !form.phrase;

  return (
    <ImportAccountFromMnemonicScreen
      form={form}
      submitDisabled={submitDisabled}
      onChange={handleChange}
      onSubmit={handleSubmit}
    />
  );
}
