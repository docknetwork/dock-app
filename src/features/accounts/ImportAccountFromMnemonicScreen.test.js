import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {ImportAccountFromMnemonicScreen} from './ImportAccountFromMnemonicScreen';

const mockStore = configureMockStore();

describe('ImportAccountFromMnemonicScreen', () => {
  it('should render correctly', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(
      <ImportAccountFromMnemonicScreen
        onChange={jest.fn()}
        form={{
          phrase: 'test',
          _errors: {
            
          }
        }}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
