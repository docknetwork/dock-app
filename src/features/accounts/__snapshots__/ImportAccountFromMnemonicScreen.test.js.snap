// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportAccountFromMnemonicScreen should render correctly 1`] = `
<StyledNativeComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "inlineStyle": InlineStyle {
        "rules": Array [
          "
  background-color: ",
          [Function],
          ";
  flex: 1;
",
        ],
      },
      "render": [Function],
      "shouldForwardProp": undefined,
      "styledComponentId": "StyledNativeComponent",
      "target": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
  testID="create-wallet-screen"
>
  <Styled(View)>
    <BackButton />
  </Styled(View)>
  <Styled(ScrollView)
    marginLeft={26}
    marginRight={26}
  >
    <Typography
      color="#fff"
      fontFamily="Montserrat"
      fontSize={24}
      fontWeight="600"
      lineHeight={32}
      marginTop={52}
    >
      Account recovery phrase
    </Typography>
    <Typography
      fontSize={16}
      fontWeight="400"
      lineHeight={24}
      marginTop={12}
    >
      This is the 12 word phrase you were given when you created your account.
    </Typography>
    <Memo(ForwardRef(Box))>
      <Memo(ForwardRef(FormControl))>
        <Memo(ForwardRef(Stack))
          mt={7}
        >
          <Memo(ForwardRef(FormControlLabel))>
            Enter account recovery phrase
          </Memo(ForwardRef(FormControlLabel))>
          <Memo(ForwardRef(TextArea))
            autoCapitalize="none"
            placeholder=""
            value="test"
          />
          <Memo(ForwardRef(FormControlErrorMessage))>
            Invalid recovery phrase
          </Memo(ForwardRef(FormControlErrorMessage))>
        </Memo(ForwardRef(Stack))>
      </Memo(ForwardRef(FormControl))>
    </Memo(ForwardRef(Box))>
    <AccountAdvancedOptions
      form={
        Object {
          "_errors": Object {},
          "phrase": "test",
        }
      }
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              "phrase",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    />
  </Styled(ScrollView)>
  <Styled(Box)
    marginBottom={10}
    marginLeft={26}
    marginRight={26}
  >
    <LoadingButton
      full={true}
      testID="next-btn"
    >
      Next
    </LoadingButton>
  </Styled(Box)>
</StyledNativeComponent>
`;
